### PR TITLE
DEX-88 Market orders introduced

### PR DESCRIPTION
**Pull request by DenisSerebryansky**

### CODE REVIEW

File: dex/src/main/scala/com/wavesplatform/dex/AddressActor.scala [hunk 1]
```diff
@@ -85,43 +88,48 @@ class AddressActor(
...
-      val batchCancelFutures = for {
-        lo <- activeOrders.values
-        if maybePair.forall(_ == lo.order.assetPair)
-      } yield {
-        val id = lo.order.id()
-        val f  = pendingCancellation.get(id).fold(storeCanceled(lo.order.assetPair, id))(_.future)
-        f.map(id -> _)
-      }
+      val batchCancelFutures =
+        for { ao <- activeOrders.values.filter { _.isLimit } if maybePair.forall(_ == ao.order.assetPair) } yield {
```
Comment (vsuharnikov) [inline dex/src/main/scala/com/wavesplatform/dex/AddressActor.scala]: Minor:
```
for { ao <- activeOrders.values.filter { _.isLimit } if maybePair.forall(_ == ao.order.assetPair) }
// vs
for { ao <- activeOrders.values if ao.isLimit && maybePair.forall(_ == ao.order.assetPair) }
```
?

File: dex/src/main/scala/com/wavesplatform/dex/AddressActor.scala [hunk 2]
```diff
@@ -301,22 +309,26 @@ class AddressActor(
...
-  case object GetReservedBalance                                          extends Command
-  case class PlaceOrder(order: Order) extends Command {
-    override lazy val toString = s"PlaceOrder(${order.id()},${order.sender},${order.assetPair},${order.orderType},${order.price},${order.amount})"
+  case class GetOrderStatus(orderId: ByteStr)                                     extends Command
+  case class GetOrdersStatuses(assetPair: Option[AssetPair], onlyActive: Boolean) extends Command
+  case class GetTradableBalance(assetPair: AssetPair)                             extends Command
+  case object GetReservedBalance                                                  extends Command
+
+  case class PlaceOrder(order: Order, isMarketOrder: Boolean = false) extends Command {
+    override lazy val toString = s"${if (isMarketOrder) "PlaceMarketOrder" else "PlaceOrder"}(${order
```
Comment (vsuharnikov) [inline dex/src/main/scala/com/wavesplatform/dex/AddressActor.scala]: May be something like this? 
```
s"PlaceOrder(${if (isMarket) "market," else ""}
```
It's easy to forget that PlaceOrder is actually "PlaceOrder" and "PlaceMarketOrder".

File: dex/src/main/scala/com/wavesplatform/dex/AddressDirectory.scala [hunk 3]
```diff
@@ -49,27 +50,31 @@ class AddressDirectory(spendableBalanceChanged: Observable[(Address, Asset)],
...
-      if (counter.order.sender != submitted.order.sender) {
-        forward(counter.order.sender, e)
+      if (counter.order.sender != submitted.order.sender) forward(counter.order.sender, e)
+
+      lazy val isFirstExecution  = submitted.amount == submitted.order.amount
+      lazy val isSubmittedFilled = e.submittedRemainingAmount == 0
+
+      (submitted.isMarket, isFirstExecution, isSubmittedFilled) match {
+        case (true, true, _)     => historyRouter foreach { _ ! SaveOrder(submitted, timestamp) }
+        case (false, true, true) => historyRouter foreach { _ ! SaveOrder(submitted, timestamp) }
```
Comment (vsuharnikov) [inline dex/src/main/scala/com/wavesplatform/dex/AddressDirectory.scala]: Should `isSubmittedFilled` affect this?

File: dex/src/main/scala/com/wavesplatform/dex/error/MatcherError.scala [hunk 4]
```diff
@@ -328,6 +328,26 @@ case class OrderInvalidPrice(ord: Order, prcSettings: OrderRestrictionsSettings)
...
+
+case class InvalidMarketOrderPrice(mo: Order)
+    extends MatcherError(
+      marketOrder,
+      price,
+      outOfBound,
+      if (mo.orderType == OrderType.BUY)
+        e"""Price of the buy market order
+         |(${'orderPrice -> mo.price}
+         |${'spendAsset -> AssetPair.assetIdStr(mo.getSpendAssetId)})
```
Comment (vsuharnikov) [inline dex/src/main/scala/com/wavesplatform/dex/error/MatcherError.scala]: Use `Price(mo.assetPair, mo.price)` here. You can see the usage in `DeviantOrderPrice`.

File: dex/src/main/scala/com/wavesplatform/dex/error/MatcherError.scala [hunk 5]
```diff
@@ -328,6 +328,26 @@ case class OrderInvalidPrice(ord: Order, prcSettings: OrderRestrictionsSettings)
...
+         |is too low for its full execution with the current market state"""
+      else
+        e"""Price of the sell market order
+         |(${'orderPrice -> mo.price}
+         |${'spendAsset -> AssetPair.assetIdStr(mo.getSpendAssetId)})
```
Comment (vsuharnikov) [inline dex/src/main/scala/com/wavesplatform/dex/error/MatcherError.scala]: Same

File: dex/src/main/scala/com/wavesplatform/dex/AddressActor.scala (hunk 1 above)
Comment (DenisSerebryansky) [reply dex/src/main/scala/com/wavesplatform/dex/AddressActor.scala]: Fixed

File: dex-it/src/test/scala/com/wavesplatform/it/sync/MarketOrderTestSuite.scala [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+      node.reservedBalance(alice) shouldBe empty
+      node.orderBook(ethWavesPair).asks shouldBe empty
+      node.orderBook(ethWavesPair).bids.size shouldBe 1
+      node.orderBook(ethWavesPair).bids.head.amount shouldBe 1 * amount
+      node.orderBook(ethWavesPair).bids.head.price shouldBe goodPrice(BUY)
+    }
+  }
+
+  "Market order should be executed correctly when available for spending < required by spendable asset" in {
+    pending
```
[the remark this answers is not in the document]
Comment (DenisSerebryansky) [reply dex-it/src/test/scala/com/wavesplatform/it/sync/MarketOrderTestSuite.scala]: Job for testers ._.

File: dex/src/main/scala/com/wavesplatform/dex/AddressActor.scala (hunk 2 above)
Comment (DenisSerebryansky) [reply dex/src/main/scala/com/wavesplatform/dex/AddressActor.scala]: Do you suggest to split that command into 2 (`PlaceLimitOrder` and `PlaceMarketOrder`)?

File: dex/src/main/scala/com/wavesplatform/dex/AddressDirectory.scala (hunk 3 above)
Comment (DenisSerebryansky) [reply dex/src/main/scala/com/wavesplatform/dex/AddressDirectory.scala]: Yes it sould

File: dex/src/main/scala/com/wavesplatform/dex/error/MatcherError.scala (hunk 4 above)
Comment (DenisSerebryansky) [reply dex/src/main/scala/com/wavesplatform/dex/error/MatcherError.scala]: Done

File: dex/src/main/scala/com/wavesplatform/dex/error/MatcherError.scala (hunk 5 above)
Comment (DenisSerebryansky) [reply dex/src/main/scala/com/wavesplatform/dex/error/MatcherError.scala]: Done

**Pull request merged**